[Enhancement]: Add interface `service_policy` parameter

acch opened the pull request:
Adds ability to configure `service_policy` parameter of IP Interfaces.

Closes #306.

Acceptance tests pass:
```shell
$ TF_ACC=1 go test ./internal/provider/networking/network_ip_interface_resource_test.go -v
=== RUN   TestAccNetworkIpInterfaceResource
--- PASS: TestAccNetworkIpInterfaceResource (4.44s)
PASS
ok      command-line-arguments  4.449s
```
```shell
$ TF_ACC=1 go test ./internal/provider/networking/network_ip_interface_resource_alias_test.go -v
=== RUN   TestAccNetworkIpInterfaceResourceAlias
--- PASS: TestAccNetworkIpInterfaceResourceAlias (3.93s)
PASS
ok      command-line-arguments  3.935s
```

Example Terraform Configuration:
```terraform
// sample IP Interface with service_policy parameter
resource "netapp-ontap_networking_ip_interface" "example" {
  cx_profile_name = ...
  name            = ...
  svm_name        = ...
  ...
  service_policy   = "default-management"
}
```

acch committed:
feat: add `service_policy` parameter to ip_interface(s) data source

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: add `service_policy` parameter to ip_interface resource

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: add acceptance tests for ip_interface resource `service_policy` parameter

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
fix: remove duplicate imports

Signed-off-by: Achim Christ <achim.christ@sva.de>

chuyich commented on internal/provider/networking/network_ip_interface_resource.go:
```diff
@@ -121,6 +122,18 @@ func (r *IPInterfaceResource) Schema(ctx context.Context, req resource.SchemaReq
+				 *   "default-management" if scope is cluster and IPspace is not Cluster (not yet implemented)
+				 *   "default-cluster" if scope is cluster and IPspace is Cluster (not yet implemented)
+				 */
+				Default: stringdefault.StaticString("default-data-files"),
```
In ONTAP 9.6 API document:
If not specified in POST, the following default property values are assigned:
:
service_policy:
default-data-files if scope is svm
default-management if scope is cluster and IPspace is not Cluster
default-cluster if scope is svm and IPspace is Cluster

The default value can be found in query API. And it does not have to be set in the POST API call. So, it should not be set here.

acch replied:
Hello @chuyich, thank you very much for reviewing this PR.

I did add a default value for the following reason: without a default value, when there is no configuration supplied, Terraform will mark the attribute as unknown. Unknown attributes are constantly displayed as "(known after apply)" in plan outputs... the default value is to reduce these "(known after apply)" plan outputs.

I'm happy to remove the default value, though - just wanted to double-check that the behavior mentioned above is indeed desired in this case.

Thanks much for a quick confirmation!

chuyich replied:
Since it could be one of 3 values based on the scope, set any of them may not be correct. It's better to use this way:
https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/name_services/name_services_ldap_resource.go#L106
`				PlanModifiers: []planmodifier.String{
					stringplanmodifier.UseStateForUnknown(),
				},`
We found setting default values may cause issues in some cases. The above will cover the case you mentioned. Also after you remove the default value part, the parameter handling needs to be adjusted accordingly. You may use the above case as a reference.

acch replied:
Understood. I've removed the default value, and added logic for fetching the value computed by ONTAP API instead.

chuyich reviewed (changes requested):
In resource, the field service_policy should not be set default value at the beginning. Based on the API (9.6) document, this field does not have to be set in the POST API call. Please make change accordingly.

carchi8py reviewed (changes requested):
Missing updates to the documentation pages for the new options.

acch committed:
fix: remove default value for `service_policy`parameter

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
chore: add `service_policy` parameter to attribute list

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch commented:
Thanks @carchi8py and @chuyich for your feedback!

I've removed the default value for `service_policy`, and added logic for fetching the value computed by ONTAP API instead. I've also updated the documentation for the new parameter, and verified that acceptance tests (still) pass.
Please let me know if this is acceptable, or what is missing.

Thanks much!

acch commented:
@carchi8py Are we good to merge this PR?

Please let me know what I need to do exactly to have the PR added to the provider. My understanding is that I've addressed all requested changes... please let me know what's missing.

Many thanks!

suhasbshekar commented:
Can you update CHANGELOG.md file?

Also your branch is behind the integration/main, we added few more checks to the pipeline as it was missing here, since your PR is from a fork of this repo, so could you pull the new changes into your branch and update the PR? 

Just for our reference, since we have disabled ACC tests as our testing profiles are down, acc tests are passing for this PR when tested locally.

acch committed:
chore: update changelog

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch commented:
@suhasbshekar Thanks for your feedback.

- I've added an entry to CHANGELOG.md
- I've rebased my branch to integration/main (once again)

Acc tests are passing on my local test system (ONTAP Select running 9.15.1P4):

```shell
$ date && TF_ACC=1 go test ./internal/provider/networking/network_ip_interface_resource_test.go -v && date
Tue Feb  4 08:48:40 UTC 2025
=== RUN   TestAccNetworkIpInterfaceResource
--- PASS: TestAccNetworkIpInterfaceResource (4.22s)
PASS
ok      command-line-arguments  4.233s
Tue Feb  4 08:48:45 UTC 2025
```

Please let me know what else is needed.
Many thanks!

suhasbshekar commented:
Please resolve the conflicts.

acch commented:
@suhasbshekar I've rebased the PR once again.

Can you please let me know when you'll be able to merge these changes?

suhasbshekar commented:
@acch I have merged one PR, please keep an eye on your PRs, as there will be conflicts once any PR is merged. For now please resolve the conflicts of this PR

acch commented:
@suhasbshekar Sure, resolved the conflict in this PR.
